### PR TITLE
arch/arm/*/up_assert.c : Show running func and its pid when asserted …

### PR DESCRIPTION
**Pull request by jeongchanKim**
…from workq

When asserted from workqueue, it is from running function, not workqueue itself.
So print the log which thread delegated the work into the workqueue.

Signed-off-by: jeongchanKim <jc_.kim@samsung.com>

### CODE REVIEW

File: os/wqueue/Kconfig [hunk 1]
```diff
@@ -177,3 +177,11 @@ config SCHED_USRWORKSTACKSIZE
 
 endif # SCHED_USRWORK
 endif # BUILD_PROTECTED || BUILD_KERNEL
+
+config DEBUG_WORKQUEUE
+	bool "Show running func, pid when asserted"
+	default n
+	---help---
```
Comment (sunghan-chang) [inline os/wqueue/Kconfig:184]: Let's make a dependency with `SCHED_WORKQUEUE`

File: os/arch/arm/src/armv8-m/up_assert.c [hunk 2]
```diff
@@ -77,6 +77,10 @@
 #include <stdlib.h>
 #include <assert.h>
 #include <debug.h>
+#if defined(CONFIG_SCHED_WORKQUEUE) && defined(CONFIG_DEBUG_WORKQUEUE)
```
Comment (sunghan-chang) [inline os/arch/arm/src/armv8-m/up_assert.c:80]: after making a dependency, let's use `CONFIG_DEBUG_WORKQUEUE` only.

File: os/arch/arm/src/armv8-m/up_assert.c [hunk 3]
```diff
@@ -479,6 +483,22 @@ void up_assert(const uint8_t *filename, int lineno)
...
+	if (strncmp(this_task()->name, HPWORKNAME) == 0) {
+		workq = (struct wqueue_s *)get_hpwork();	
+	}
+#endif
+#ifdef CONFIG_SCHED_LPWORK
+	if (strncmp(this_task()->name, LPWORKNAME) == 0) {
+		workq = (struct wqueue_s *)get_hpwork();
+	}
+#endif
+	lldbg("Running work function is %x, from pid %d.\n", workq->q.head, workq->worker[0].pid);
```
Comment (sunghan-chang) [inline os/arch/arm/src/armv8-m/up_assert.c:499]: 1. need checking user workqueue.
2. Even if nothing is matched, this line is executed. This should be printed only when wq is asserted.

File: os/arch/arm/src/armv8-m/up_assert.c [hunk 4]
```diff
@@ -479,6 +483,22 @@ void up_assert(const uint8_t *filename, int lineno)
 #else
 	lldbg("Assertion failed at file:%s line: %d\n", filename, lineno);
 #endif
+
+#if defined(CONFIG_SCHED_WORKQUEUE) && defined(CONFIG_DEBUG_WORKQUEUE)
+	struct wqueue_s *workq;
+#ifdef CONFIG_SCHED_HPWORK
+	if (strncmp(this_task()->name, HPWORKNAME) == 0) {
```
Comment (sunghan-chang) [inline os/arch/arm/src/armv8-m/up_assert.c:490]: there is no size.

File: os/wqueue/Kconfig (hunk 1 above)
Comment (jeongchanKim) [reply os/wqueue/Kconfig:185]: Added. Thanks.

File: os/arch/arm/src/armv7-m/up_assert.c [hunk 5]
```diff
@@ -462,6 +470,34 @@ void up_assert(const uint8_t *filename, int lineno)
...
+	}
+#endif
+#ifdef CONFIG_SCHED_LPWORK
+	if (strncmp(this_task()->name, LPWORKNAME, sizeof(HPWORKNAME)) == 0) {
+		workq = (struct wqueue_s *)get_hpwork();
+		is_asserted_workq = true;
+	}
+#endif
+
+#ifdef CONFIG_SCHED_LPWORK
```
Comment (sunghan-chang) [inline os/arch/arm/src/armv7-m/up_assert.c:490]: typo for LPWORK

File: os/arch/arm/src/armv7-m/up_assert.c [hunk 6]
```diff
@@ -462,6 +470,34 @@ void up_assert(const uint8_t *filename, int lineno)
...
+	if (strncmp(this_task()->name, "uwork", sizeof("uwork")) == 0) {
+		workq = (struct wqueue_s *)get_usrwork();
+		is_asserted_workq = true;
+	}
+#endif
+	if (is_asserted_workq == true) {
+		lldbg("Running work function is %x, from pid %d.\n", workq->q.head, workq->worker[0].pid);
```
Comment (sunghan-chang) [inline os/arch/arm/src/armv7-m/up_assert.c:497]: `work->q.head` is not the function. 
It is `(work->q.head)->worker`. could you do double check?

File: os/wqueue/work_process.c [hunk 7]
```diff
@@ -185,6 +181,22 @@ void work_process(FAR struct wqueue_s *wqueue, int wndx)
 #endif
 				worker(arg);
 
+#if defined(CONFIG_SCHED_USRWORK) && !defined(__KERNEL__)
+	while (work_lock() < 0);
+#else
+	flags = irqsave();
```
Comment (sunghan-chang) [inline os/wqueue/work_process.c:187]: Please check indentation.

File: os/arch/arm/src/armv7-m/up_assert.c [hunk 8]
```diff
@@ -462,6 +470,35 @@ void up_assert(const uint8_t *filename, int lineno)
...
+	if (strncmp(this_task()->name, "uwork", sizeof("uwork")) == 0) {
+		workq = (struct wqueue_s *)get_usrwork();
+		is_asserted_workq = true;
+	}
+#endif
+
+	if (is_asserted_workq == true) {
+		lldbg("Running work function is %x.\n", ((struct work_s *)workq->q.head)->worker);
+	}
+#endif /* defined(CONFIG_SCHED_WORKQUEUE) && defined(CONFIG_DEBUG_WORKQUEUE) */
```
Comment (abhishek-samsung) [inline os/arch/arm/src/armv7-m/up_assert.c:500]: please remove defined(CONFIG_SCHED_WORKQUEUE) as it is no longer checked.

File: os/arch/arm/src/armv7-r/arm_assert.c [hunk 9]
```diff
@@ -950,6 +954,35 @@ void up_assert(const uint8_t *filename, int lineno)
...
+	if (strncmp(this_task()->name, "uwork", sizeof("uwork")) == 0) {
+		workq = (struct wqueue_s *)get_usrwork();
+		is_asserted_workq = true;
+	}
+#endif
+
+	if (is_asserted_workq == true) {
+		lldbg("Running work function is %x.\n", ((struct work_s *)workq->q.head)->worker);
+	}
+#endif /* defined(CONFIG_SCHED_WORKQUEUE) && defined(CONFIG_DEBUG_WORKQUEUE) */
```
Comment (abhishek-samsung) [inline os/arch/arm/src/armv7-r/arm_assert.c:983]: same here.

File: os/arch/arm/src/armv8-m/up_assert.c [hunk 10]
```diff
@@ -479,6 +485,35 @@ void up_assert(const uint8_t *filename, int lineno)
...
+	if (strncmp(this_task()->name, "uwork", sizeof("uwork")) == 0) {
+		workq = (struct wqueue_s *)get_usrwork();
+		is_asserted_workq = true;
+	}
+#endif
+
+	if (is_asserted_workq == true) {
+		lldbg("Running work function is %x.\n", ((struct work_s *)workq->q.head)->worker);
+	}
+#endif /* defined(CONFIG_SCHED_WORKQUEUE) && defined(CONFIG_DEBUG_WORKQUEUE) */
```
Comment (abhishek-samsung) [inline os/arch/arm/src/armv8-m/up_assert.c:514]: same here

File: os/wqueue/wqueue.h [hunk 11]
```diff
@@ -274,5 +274,15 @@ void work_process(FAR struct wqueue_s *wqueue, int wdx);
...
 int work_qsignal(pid_t pid);
 
+/****************************************************************************
+ * Name: work_qsignal
+ *
+ * Description:
+ *   Get the last running worker
+ ****************************************************************************/
+#ifdef CONFIG_DEBUG_WORKQUEUE
+worker_t work_get_current(void);
```
Comment (sunghan-chang) [inline os/wqueue/wqueue.h:284]: Let's move into os/include/tinyara/wqueue.h

File: os/arch/arm/src/armv8-m/up_assert.c [hunk 12]
```diff
@@ -479,6 +485,35 @@ void up_assert(const uint8_t *filename, int lineno)
 #else
 	lldbg("Assertion failed at file:%s line: %d\n", filename, lineno);
 #endif
+
+#ifdef CONFIG_DEBUG_WORKQUEUE
+	struct wqueue_s *workq;
+	bool is_asserted_workq = false;
+#ifdef CONFIG_SCHED_HPWORK
+	if (strncmp(this_task()->name, HPWORKNAME, sizeof(HPWORKNAME)) == 0) {
+		workq = (struct wqueue_s *)get_hpwork();
```
Comment (sunghan-chang) [inline os/arch/arm/src/armv8-m/up_assert.c:494]: unnecessary.

File: os/arch/arm/src/armv8-m/up_assert.c [hunk 13]
```diff
@@ -479,6 +485,35 @@ void up_assert(const uint8_t *filename, int lineno)
 #else
 	lldbg("Assertion failed at file:%s line: %d\n", filename, lineno);
 #endif
+
+#ifdef CONFIG_DEBUG_WORKQUEUE
+	struct wqueue_s *workq;
+	bool is_asserted_workq = false;
```
Comment (sunghan-chang) [inline os/arch/arm/src/armv8-m/up_assert.c:491]: How about just compring in a single if without this new variable?
```
if (!strncmp(hpwork) || !strncmp(lpwork) || !strncmp(usrwork)) {
    lldbg("Running work is 0x%x\n", work_get_current());
```

File: os/arch/arm/src/armv8-m/up_assert.c [hunk 14]
```diff
@@ -479,6 +483,13 @@ void up_assert(const uint8_t *filename, int lineno)
 #else
 	lldbg("Assertion failed at file:%s line: %d\n", filename, lineno);
 #endif
+
+#ifdef CONFIG_DEBUG_WORKQUEUE
+	if (IS_HPWORK || IS_LPWORK) {
+		lldbg("Running work function is %x.\n", work_get_current());
```
Comment (sunghan-chang) [inline os/arch/arm/src/armv8-m/up_assert.c:489]: Is there anyway to know which wq is?

File: os/wqueue/work_process.c [hunk 15]
```diff
@@ -182,6 +190,9 @@ void work_process(FAR struct wqueue_s *wqueue, int wndx)
 				work_unlock();
 #else
 				irqrestore(flags);
+#endif
+#ifdef CONFIG_DEBUG_WORKQUEUE
```
Comment (sunghan-chang) [inline os/wqueue/work_process.c:194]: Need __kernel__ to exclude user workqueue

File: os/arch/arm/src/armv8-m/up_assert.c (hunk 14 above)
Comment (jeongchanKim) [reply os/arch/arm/src/armv8-m/up_assert.c:489]: In assert log, there is already task name.
I can print the name again, but it is duplicated I think.

**Comment by sunghan-chang** [comment]
@jeongchanKim Could you resolve build break?
```
CC:  kwqueue/kwork_hpthread.c
kwqueue/kwork_hpthread.c: In function 'work_hpstart':
kwqueue/kwork_hpthread.c:187:22: error: 'HPWORKNAME' undeclared (first use in this function)
  pid = kernel_thread(HPWORKNAME, CONFIG_SCHED_HPWORKPRIORITY, CONFIG_SCHED_HPWORKSTACKSIZE, (main_t)work_hpthread, (FAR char *const *)NULL);
                      ^
kwqueue/kwork_hpthread.c:187:22: note: each undeclared identifier is reported only once for each function it appears in
Makefile:110: recipe for target 'bin/kwork_hpthread.o' failed
```

File: os/include/tinyara/wqueue.h [hunk 16]
```diff
@@ -488,6 +491,19 @@ void lpwork_boostpriority(uint8_t reqprio);
...
+ *
+ * Description:
+ *   Get the current running worker
+ ****************************************************************************/
+#if defined(CONFIG_DEBUG_WORKQUEUE) && defined(__KERNEL__)
+worker_t work_get_current(void);
+
+#define IS_HPWORK (strncmp(this_task()->name, HPWORKNAME, sizeof(HPWORKNAME)) == 0)
+#define IS_LPWORK (strncmp(this_task()->name, LPWORKNAME, sizeof(HPWORKNAME)) == 0)
+#endif
```
Comment (sunghan-chang) [inline os/include/tinyara/wqueue.h:505]: Even hp and lp workqueue config can be controlled separately, here both are used without the conditional.
It causes build break. To keep code simple, I suggest below.
```
#if defined(CONFIG_DEBUG_WORKQUEUE) && defined(__KERNEL__)
worker_t work_get_current(void);

#ifdef CONFIG_SCHED_HPWORK
#define IS_HPWORK (strncmp(this_task()->name, HPWORKNAME, sizeof(HPWORKNAME)) == 0)
#else
#define IS_HPWORK (false)
#endif
#ifdef CONFIG_SCHED_LPWORK
#define IS_LPWORK (strncmp(this_task()->name, LPWORKNAME, sizeof(HPWORKNAME)) == 0)
#else
#define IS_LPWORK (false)
#endif
#endif
```

File: os/arch/arm/src/armv8-m/up_assert.c [hunk 17]
```diff
@@ -100,13 +100,17 @@
...
-#include <unistd.h>
 #include <stdbool.h>
+#include <unistd.h>
 #include <queue.h>
 #include <tinyara/wdog.h>
 #include "semaphore/semaphore.h"
 #include "binary_manager/binary_manager.h"
 #endif
+#if defined(CONFIG_DEBUG_WORKQUEUE) && defined(__KERNEL__)
+#include <string.h>
```
Comment (sunghan-chang) [inline os/arch/arm/src/armv8-m/up_assert.c:111]: Here is no string function. It's in wqueue.h. Let's move this including into the header.

File: os/include/tinyara/wqueue.h (hunk 16 above)
Comment (jeongchanKim) [reply os/include/tinyara/wqueue.h]: Modified. Thanks.

**Review by sunghan-chang — APPROVED** [review]

**Pull request merged**